### PR TITLE
feat(gatsby): Enable shadowing of page template queries

### LINKED ISSUE #16927: Shadowing templates doesn't allow you to override the page query
## Description

Shadowing `gatsby-theme-blog-core/src/templates/post-query.js` lets you shadow the component, but doesn't let you shadow the page query.

### Steps to reproduce

```
gatsby new shadow-templates https://github.com/gatsbyjs/gatsby-starter-blog-theme
```

shadow `gatsby-theme-blog-core/src/templates/post-query.js` and use whatever query you want (I used `site { siteMetadata { author }}`). You'll get a "this query is not used" warning pointing at the shadowed file.

### Expected result

the "this query is not used" warning should point at the node_module file, `gatsby-theme-blog-core/src/templates/post-query.js` since that query would be no longer used.

### Actual result

warns on the shadowed file

### PR DESCRIPTION
**Pull request by ChristopherBiscardi**
## Description

fixes #16927 

Enables shadowing for createPage calls with respect to template queries. This is a shadowing algorithm application *outside* of webpack and as such needed explicit support.

### CODE REVIEW

File: packages/gatsby/src/redux/actions/public.js [hunk 1]
```diff
@@ -226,6 +227,45 @@ ${reservedFields.map(f => `  * "${f}"`).join(`\n`)}
     }
   }
 
+  // TODO: start shadowing implementation
+  const shadowingPlugin = new GatsbyThemeComponentShadowingResolverPlugin({
+    // TODO: extensions needs to be sourced from store somehow
+    extensions: [`.js`],
```
[the remark this answers is not in the document]
Comment (ChristopherBiscardi) [reply packages/gatsby/src/redux/actions/public.js]: also move this logic to separate file

**Comment by ChristopherBiscardi** [comment]
@gatsbyjs/core this is rebased and ready for review

**Comment by pieh** [comment]
Thanks! Sorry for the delay - code LGTM

I do have worries about adding/removing shadowed page templates while `gatsby develop` is running, but this part is already problematic with webpack plugin so I don't want this to block the PR.

I will add test cases for e2e themes tests

File: e2e-tests/themes/development-runtime/package.json [hunk 2]
```diff
@@ -4,7 +4,7 @@
   "version": "1.0.0",
   "author": "Sidhartha Chatterjee <sid@gatsbyjs.com>",
   "dependencies": {
-    "gatsby": "^2.13.14",
+    "gatsby": "gatsby-dev",
```
Comment (pieh) [inline e2e-tests/themes/development-runtime/package.json]: oooopsie

Comment (sidharthachatterjee) [reply e2e-tests/themes/development-runtime/package.json]: 😂

**Comment by pieh** [comment]
Published `gatsby@2.18.1`

**Comment by laurenskling** [comment]
Can confirm this is working for my project <3 thanks guys

**Comment by jsg2021** [comment]
I hit an edge case. Create a site from `@lekoarts/gatsby-theme-minimal-blog` and attempt to shadow the `template/homepage-query.tsx` with a local `src/@lekoarts/gatsby-theme-minimal-blog-core/templates/homepage-query.js`. It won't shadow over the in-theme file. Rename it with the same extension...it works.

the documents for `createPage` tells you to use an absolute path to the component. So, the theme author didn't do anything wrong. (per docs) The component will always have an extension, so the algorithm resolving the shadowed component will not allow extension overriding. :(

**Comment by Kalmuraee** [comment]
I have the same problem shadowing `data.query.js` in tsx template :(

**Pull request merged**